proposals/type-aliases.md

```diff
@@ -131,13 +159,46 @@ typealias Encoded<E> = ByteArray         // Warning: generic type parameter E is
 >
 > fun processIsoEncoded(data: Encoded<Iso>) {
 >   // Both 'Encoded<Iso>' and 'Encoded<Utf8>' are expanded to 'ByteArray'.
->   // So, regardless of the design intent, the following call is ok:
 >   processUtf8Encoded(data)
 > }
 >```
 >
 > However, if we prohibit such type aliases, it creates unnecessary long-term commitment for generic type aliases.
-> Since it is not an error for generic classes,
+> NB this is not an error for generic classes.
+
+Type arguments of a generic type alias should be well-formed types (type projections).
+```
+interface NN<T : Any>
+typealias Predicate<T> = (T) -> Boolean
+
+typealias E1 = Predicate<NN<Int, Int>>          // Error: wrong number of type arguments
+typealias E2 = Predicate<NN<Int?>>              // Error: upper bound violated
+```
+
+Type arguments of generic type aliases are substituted syntactically.
+```
+typealias Dictionary<V> = Map<String, V>
+typealias Predicate<T> = (T) -> Boolean         // = Function1<T, Boolean>
+typealias TMap<T> = Map<T, T>
+typealias Array2D<T> = Array<Array<T>>
+
+typealias T1 = Dictionary<*>                    // Map<String, *>
+typealias T2 = Predicate<*>                     // Ok (but marginally useful): Function1<*, Boolean>
+typealias T3 = TMap<*>                          // Map<*, *>
+typealias T4 = Array2D<out Number>              // Array<Array<out Number>>
+```
+> Note that type aliases are not types, but rather macros on types.
+> E.g.:
+>
+>```
+> typealias TMap<T> = Map<T, T>
+>
+> fun <T> foo(m: TMap<T>) {}         // fun <T> (m: Map<T, T>)
+>
+> fun bar(m: TMap<*>) {              // fun     (m: Map<*, *>)
+>     foo(m)                         // Error: cannot infer type parameter T
+> }
+>```
```

Comment:
I think we should add an example about `Array2D<String>` being not assignable to `Array2D<*>`